src/lmdb.ml

```diff
@@ -233,39 +236,72 @@ module Values = struct
   module Int : S with type t = int = struct
     type t = int
     let default_flags = Flags.none
-    let int_size = Size_t.of_int (sizeof camlint)
-    let write i =
-      let cint = allocate camlint i in
-      let v = make mdb_val in
-      alive_while cint v;
-      setf v mv_size int_size ;
-      setf v mv_data (to_voidp cint) ;
-      addr v
-    let read v =
-      !@(from_voidp camlint @@ getf !@v mv_data)
+    let int_size = sizeof camlint
+
+    let bigint =
+      begin fun alloc x ->
+        let a = alloc int_size in
+        let rec loop i x =
+          Bigarray.Array1.set a i (char_of_int (x land 0xff));
+          if i > 0
+          then loop (i-1) (x lsr 8)
+        in loop (int_size-1) x;
+        a
+      end,
+      begin fun a ->
+        let rec loop i x =
+          if i < int_size
+          then loop (i+1) (x lsl 8 lor int_of_char (Bigarray.Array1.get a i))
+          else x
+        in loop 0 0
+      end
+
+    let littleint =
+      begin fun alloc x ->
+        let a = alloc int_size in
+        let rec loop i x =
+          if i < int_size
+          then begin
+            Bigarray.Array1.set a i (char_of_int (x land 0xff));
+            loop (i+1) (x lsr 8)
+          end
+        in loop 0 x;
+        a
+      end,
+      begin fun a ->
+        let rec loop i x =
+          if i >= 0
+          then loop (i-1) (x lsl 8 lor int_of_char (Bigarray.Array1.get a i))
+          else x
+        in loop (int_size-1) 0
+      end
+
+    let write, read =
+      if Sys.big_endian
+      then fst bigint alloc_buffer, snd bigint
+      else fst littleint alloc_buffer, snd littleint
```

Comment:
I think the tuple game you play here is pretty unnecessary, and doesn't help in any way. Just create toplevel read and write functions. It will even make the code shorter by avoiding all those begin/end/fun.

Reply:
After reading the next commit, I understand why you do that, but I prefer a more regular programming style.

Also, I would suggest checking which pattern leads to better inlining.

Reply:
The "tuple game" is copy-pasted from the polymorphic interface, where it is used in this way:
``` ocaml
module Conv : sig
  type 'a t = ((int -> buffer) -> 'a -> buffer) * (buffer -> 'a)

  val bigint : int t
  val littleint : int t
  val hostint : int t
  val string : string t
  val buffer : buffer t
end
```
For the polymorphic interface inlining will go out the window anyway. The performance penalty is 2% for the ``int`` case. It will probably be smaller for bigger keys / values. I believe this is negligible.
I think performance tuning like inlining should be postponed until the interface is settled.
So could we please leave this as-is for the time being. Or include Bigstringaf right away?